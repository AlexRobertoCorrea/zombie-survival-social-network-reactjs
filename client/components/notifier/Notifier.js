import React, { Component } from 'react';
import { observer } from 'mobx-react';
import { observable } from 'mobx';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import CheckCircleIcon from '@material-ui/icons/CheckCircle';
import CloseIcon from '@material-ui/icons/Close';
import ErrorIcon from '@material-ui/icons/Error';
import IconButton from '@material-ui/core/IconButton';
import InfoIcon from '@material-ui/icons/Info';
import Snackbar from '@material-ui/core/Snackbar';
import SnackbarContent from '@material-ui/core/SnackbarContent';
import green from '@material-ui/core/colors/green';
import { withStyles } from '@material-ui/core/styles';

const variantIcon = {
  success: CheckCircleIcon,
  error: ErrorIcon,
  info: InfoIcon
};

const styles = theme => ({
  success: {
    backgroundColor: green[600]
  },
  error: {
    backgroundColor: theme.palette.error.dark
  },
  info: {
    backgroundColor: theme.palette.primary.dark
  },
  icon: {
    fontSize: 20
  },
  iconVariant: {
    opacity: 0.9,
    marginRight: theme.spacing.unit
  },
  message: {
    display: 'flex',
    alignItems: 'center'
  }
});

@observer class Notifier extends Component {
  handleClose = (event, reason) => {
    if (reason === 'clickaway') {
      return;
    }

    this.open = false;
  };

  update(data) {
    this.variant = data.variant;
    this.message = data.message;
    this.open = true;
  }
  
  @observable open = false;
  @observable variant = null;
  @observable message = null;
  
  renderIcon() {
    if (!this.variant) {
      return;
    }
  
    const { classes } = this.props;
    const Icon = variantIcon[this.variant];
    
    return <Icon className={classNames(classes.icon, classes.iconVariant)} />;
  }
  
  render() {
    const { classes } = this.props;
    
    return (
      <Snackbar
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'right'
        }}
        open={this.open}
        autoHideDuration={6000}
        onClose={this.handleClose}
      >
        <SnackbarContent
          className={classes[this.variant]}
          aria-describedby="client-snackbar"
          message={
            <span id="client-snackbar" className={classes.message}>
              {this.renderIcon()}
              {this.message}
            </span>
          }
          action={[
            <IconButton
              key="close"
              aria-label="Close"
              color="inherit"
              onClick={this.handleClose}
            >
              <CloseIcon />
            </IconButton>
          ]}
        />
      </Snackbar>
    );
  }
}

Notifier.propTypes = {
  classes: PropTypes.object.isRequired
};

export default withStyles(styles)(Notifier);
