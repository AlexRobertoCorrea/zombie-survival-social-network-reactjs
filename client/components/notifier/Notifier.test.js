import React from 'react';

import Notifier from './Notifier';

describe('<Notifier />', () => {
  describe('Render when open is false', () => {
    const classes = {
      message: 'success'
    };
  
    const wrapper = shallow(<Notifier classes={classes} />);
  
    wrapper.setState({
      open: false,
      message: 'message test',
      variant: 'success'
    });
    
    it('Ensure that Notifier rendered correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
  
  describe('Render when open is true and success message', () => {
    const classes = {
      message: 'success',
      success: 'success'
    };
  
    const wrapper = shallow(<Notifier classes={classes} />);
    
    wrapper.setState({
      open: true,
      message: 'message test',
      variant: 'success'
    });

    it('Ensure that Notifier rendered correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
  
  describe('Render when open is true and error message', () => {
    const classes = {
      message: 'success',
      error: 'error'
    };
    
    const wrapper = shallow(<Notifier classes={classes} />);
    
    wrapper.setState({
      open: true,
      message: 'error message',
      variant: 'error'
    });
    
    it('Ensure that Notifier rendered correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
