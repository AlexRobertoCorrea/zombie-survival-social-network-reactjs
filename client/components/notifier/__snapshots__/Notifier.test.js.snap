// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notifier /> When open is false Notifier is rendered correctly 1`] = `
<Notifier
  classes={
    Object {
      "error": "Notifier-error-02",
      "icon": "Notifier-icon-04",
      "iconVariant": "Notifier-iconVariant-05",
      "info": "Notifier-info-03",
      "message": "Notifier-message-06 success",
      "success": "Notifier-success-01",
    }
  }
/>
`;

exports[`<Notifier /> When open is true and error message Notifier is rendered correctly with error message 1`] = `
<Notifier
  classes={
    Object {
      "error": "Notifier-error-02 error",
      "icon": "Notifier-icon-04",
      "iconVariant": "Notifier-iconVariant-05",
      "info": "Notifier-info-03",
      "message": "Notifier-message-06 success",
      "success": "Notifier-success-01",
    }
  }
/>
`;

exports[`<Notifier /> When open is true and success message Notifier is rendered with success message 1`] = `
<Notifier
  classes={
    Object {
      "error": "Notifier-error-02",
      "icon": "Notifier-icon-04",
      "iconVariant": "Notifier-iconVariant-05",
      "info": "Notifier-info-03",
      "message": "Notifier-message-06 success",
      "success": "Notifier-success-01 success",
    }
  }
/>
`;
